src/Stratis.Bitcoin.Features.Miner/BlockAssembler.cs

```diff
@@ -0,0 +1,559 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+using NBitcoin;
+using Stratis.Bitcoin.Base.Deployments;
+using Stratis.Bitcoin.Features.Consensus;
+using Stratis.Bitcoin.Features.Consensus.Interfaces;
+using Stratis.Bitcoin.Features.MemoryPool;
+using Stratis.Bitcoin.Features.MemoryPool.Interfaces;
+using Stratis.Bitcoin.Mining;
+using Stratis.Bitcoin.Utilities;
+using static Stratis.Bitcoin.Features.Miner.PowBlockAssembler;
+
+namespace Stratis.Bitcoin.Features.Miner
+{
+    public abstract class BlockAssembler
+    {
+        /// <summary>
+        /// Tip of the chain that this instance will work with without touching any shared chain resources.
+        /// </summary>
+        protected ChainedHeader ChainTip;
+
+        /// <summary>Manager of the longest fully validated chain of blocks.</summary>
+        protected readonly IConsensusLoop ConsensusLoop;
+
+        /// <summary>Provider of date time functions.</summary>
+        protected readonly IDateTimeProvider DateTimeProvider;
+
+        /// <summary>Instance logger.</summary>
+        private readonly ILogger logger;
+
+        /// <summary>Transaction memory pool for managing transactions in the memory pool.</summary>
+        protected readonly ITxMempool Mempool;
+
+        /// <summary>Lock for memory pool access.</summary>
+        protected readonly MempoolSchedulerLock MempoolLock;
+
+        /// <summary>The current network.</summary>
+        protected readonly Network Network;
+
+        /// <summary>Assembler options specific to the assembler e.g. <see cref="AssemblerOptions.BlockMaxSize"/>.</summary>
+        protected AssemblerOptions Options;
+
+        private const long TicksPerMicrosecond = 10;
```

Comment:
As we are touching this file, can we get rid of this constant, it smells very much.?

Reply:
It is fine to say that it is out of the scope of this PR.

Reply:
I've removed it 👍